### PR TITLE
[8.0] Pilot submission with tokens

### PR DESCRIPTION
**Pull request by atsareg**
This PR adds possibility to submit pilots with tokens. The tokens are obtained by SiteDirectors via a standard mechanism from the TokenManager. In order to use the oidc-agent instead of the OAuth2 flow, a new OidcAgentIdProvider is added. Later on, switching to the direct use of OAuth2 providers should be transparent by changing the IdProvider option of the pilot group, e.g. lhcb_pilot.
  SiteDirector is enabled to select queues by Tag values. Queues can define Token tag to be used with tokens, otherwise with certificates. Later on, the default will be switched to Token I hope :).
  The oidc-agent is supposed to have a client created with the name of the user submitting pilots and having necessary compute scopes.    

  The following ComputingElement should be enabled with the tokens: 

- [x] HTCondorCE
- [x] ARC/ARC6
- [ ] AREX
- [ ] Cloud

BEGINRELEASENOTES

*Framework
NEW: OidcAgentIdProvider identity provider class

*WorkloadManagement
NEW: SiteDirector enabled to select queues by tags
NEW: SiteDirector sets up tokens for ComputingElements configured with the Token tag

*Resources
NEW: HTCondorComputingElement and ARC(6)ComputingElement enabled to for job operations with tokens

*Configuration
CHANGE: Registry - IdProvider can be defined either on VO or Group level

ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/FrameworkSystem/Client/TokenManagerClient.py [hunk 1]
```diff
@@ -86,5 +86,7 @@ def getToken(
                 idpObj.revokeToken(token["refresh_token"])
 
         return self.executeRPC(
-            username, userGroup, scope, audience, identityProvider, requiredTimeLeft, call="getToken"
+            userName, userGroup, scope, audience, identityProvider, requiredTimeLeft, call="getToken"
         )
+
+gTokenManager = TokenManagerClient()
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/Client/TokenManagerClient.py:93]: Is it really needed?

File: src/DIRAC/FrameworkSystem/Service/TokenManagerHandler.py [hunk 2]
```diff
@@ -241,9 +241,30 @@ def export_getToken(
...
                 idpObj.revokeToken(token["refresh_token"])
 
+        # No token in the cache, try getting
+        kwargs = {"userName": userName,
+                  "timeLeft": requiredTimeLeft,
+                  "group": userGroup,
+                  "scope": scope,
+                  "audience": audience,
+                  }
+        result = idpObj.getToken( **kwargs )
```
Comment (fstagni) [inline src/DIRAC/FrameworkSystem/Service/TokenManagerHandler.py]: seems like `idpObj.getToken` returns an `S_OK/S_ERROR` structure, while `idpObj.getGroupScopes` does not. It feels inconsistent.

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 3]
```diff
@@ -372,42 +418,79 @@ def getJobStatus(self, jobIDList):
...
         self.log.verbose("Pilot Statuses: %s " % resultDict)
         return S_OK(resultDict)
 
+    def _treatCondorHistory(self, condorHistCall, qList):
+        """concatenate clusterID and processID to get the same output as condor_q
+        until we can expect condor version 8.5.3 everywhere
+
+        :param str condorHistCall: condor_history command to run
+        :param qList: list of jobID and status from condor_q output, will be modified in this function
+        :type qList: python:list
```
Comment (fstagni) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:477]:
```suggestion
        :param list qList: list of jobID and status from condor_q output, will be modified in this function
```

File: src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py [hunk 4]
```diff
@@ -306,7 +306,7 @@ def getPilotFilesCompressedEncodedDict(pilotFiles, proxy=None):
         pilotFilesCompressedEncodedDict[os.path.basename(pf)] = pfContentEncoded
 
     if proxy is not None:
-        compressedAndEncodedProxy = base64.b64encode(bz2.compress(proxy.dumpAllToString()["Value"]))
+        compressedAndEncodedProxy = base64.b64encode(bz2.compress(proxy.dumpAllToString()["Value"].encode()))
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py:309]: Is this a bug fix? Why would we have not needed it up to now?

File: src/DIRAC/ConfigurationSystem/Client/Helpers/Resources.py [hunk 5]
```diff
@@ -284,6 +297,10 @@ def getQueues(siteList=None, ceList=None, ceTypeList=None, community=None):
                         comList = gConfig.getValue(f"/Resources/Sites/{grid}/{site}/CEs/{ce}/Queues/{queue}/VO", [])
                         if comList and community.lower() not in [cl.lower() for cl in comList]:
                             continue
+                    if tags:
+                        queueTags = gConfig.getValue(f"/Resources/Sites/{grid}/{site}/CEs/{ce}/Queues/{queue}/Tag", [])
+                        if not ceTags or not set(tags).issubset(set(queueTags)):
```
Comment (fstagni) [inline src/DIRAC/ConfigurationSystem/Client/Helpers/Resources.py]:
```suggestion
                        if not queueTags or not set(tags).issubset(set(queueTags)):
```
?

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 6]
```diff
@@ -173,6 +173,17 @@ def __writeSub(self, executable, nJobs, location, processors):
 
         executable = os.path.join(self.workingDirectory, executable)
 
+        useCredentials = ""
+        if tokenFile:
+            useCredentials = (
+                """
+use_x509userproxy = true
```
Comment (fstagni) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:180]: does this need to stay "true"?

File: src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py [hunk 7]
```diff
@@ -435,6 +442,20 @@ def submitPilots(self):
             lifetime_secs = result["Value"]
             ce.setProxy(proxy, lifetime_secs)
 
+            # Get valid token id needed
+            if "Token" in ce.ceParameters.get("Tag", []):
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py:446]: I am not sure I like this approach. Should not this based on the CEType? Or what if we do it always?

File: src/DIRAC/FrameworkSystem/Client/TokenManagerClient.py (hunk 1 above)
Comment (atsareg) [reply src/DIRAC/FrameworkSystem/Client/TokenManagerClient.py:93]: It is similar to ProxyManagerClient. It should be a global object to keep the cache of already available tokens across multiple threads for example

File: src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py (hunk 4 above)
Comment (atsareg) [reply src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py:309]: This is a bug. I think we were simply not hitting it before

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 6 above)
Comment (atsareg) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:180]: I think yes. This is the way to tell condor to pass also proxy along with the token. If we choose always to bundle proxy in, this is not needed but does not harm neither

File: src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py (hunk 7 above)
Comment (atsareg) [reply src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py:446]: This has nothing to do with the CEType. Different CETypes can use either certificates or tokens. But there can be a reasonable default. Now the default is no Token meaning using certificates

File: src/DIRAC/ConfigurationSystem/Client/Helpers/Resources.py (hunk 5 above)
Comment (atsareg) [reply src/DIRAC/ConfigurationSystem/Client/Helpers/Resources.py]: Indeed, the latest tested version of these lines was not committed, now done

File: src/DIRAC/FrameworkSystem/Service/TokenManagerHandler.py (hunk 2 above)
Comment (atsareg) [reply src/DIRAC/FrameworkSystem/Service/TokenManagerHandler.py]: The two methods are far different in complexity. getGroupScopes() would only return S_OK and never S_ERROR. In this case we usually do not return the structure

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py (hunk 6 above)
Comment (fstagni) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:180]: I think we said to always bundle the proxy (seems less error-prone).

File: src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py (hunk 7 above)
Comment (fstagni) [reply src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py:446]: Is this an information that we can get from BDII? CC @andresailer

**Pull request closed without merging**